OSDOCS-13459#Add cross-subscription support for Azure File (#92108)
openshift/openshift-docs: modules/persistent-storage-csi-azure-file-cross-sub-dynamic-provisioning-procedure.adoc

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-provisioning-procedure_{context}"]
= Dynamic provisioning across subscriptions for Azure File

.Prerequisites

> * Installed Azure cluster with the service principal or managed identity as Azure identity

Reviewer: See [above](https://github.com/openshift/openshift-docs/pull/92108/files#r2040178914).

* Access to another subscription in the tenant with the cluster  

* Logged in to the Azure CLI

.Procedure
To use Azure File dynamic provisioning across subscriptions:

> . Record the Azure identity (service principal or managed identity) by running the following applicable commands. The Azure identity is needed in the next step:

Reviewer: Do we need (service principal or managed identity) here, as it is in the prereqs?

```suggestion
. Record the Azure identity by running the following applicable commands. The Azure identity is needed in the next step:
```

+

> * If using the _service principal_ as Azure identity when installing the cluster:

Reviewer: See [above](https://github.com/openshift/openshift-docs/pull/92108/files#r2040178914).

+

> [source,terminal]
> ====
> $ sp_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)
> 
> $ az ad sp show --id ${sp_id} --query displayName --output tsv
> ====

Reviewer: ```suggestion
[source,terminal]
----
$ sp_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)

$ az ad sp show --id ${sp_id} --query displayName --output tsv
----
```

+

> * If using the _managed identity_ as Azure identity when installing the cluster:

Reviewer: See [above](https://github.com/openshift/openshift-docs/pull/92108/files#r2040178914).

+
[source,terminal]
====
$ mi_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)

$ az identity list --query "[?clientId=='${mi_id}'].{Name:name}" --output tsv
====

> . Obtain the Azure identity (service principal or managed identity) permission to access the resource group in another subscription where you want to provision the Azure File share:

Reviewer: Do we need (service principal or managed identity) here, as it is in the prereqs?

```suggestion
. Record the Azure identity by running the following applicable commands. The Azure identity is needed in the next step:
```

.. Log in to the Azure portal and on the left-hand menu, click *Resource groups*.

> .. Choose the resource group to which you want to assign a role by clicking *resource group* > *Access control (IAM)* > *Role assignments* tab to view current assignments, and then click *Add* > *Add role assignment*.

Reviewer: Is *resource group* capitalized properly?
```suggestion
.. Choose the resource group to which you want to assign a role by clicking *resource group* -> *Access control (IAM)* -> *Role assignments* tab to view current assignments, and then click *Add* > *Add role assignment*.
```

.. On the *Role* tab, choose the contributor role to assign, and then click *Next*. You can also create and choose your own role with required permission.

> .. On the *Members* tab, choose an assignee by selecting the type of assignee: “User, group, or service principal” (Or “Managed identity”), click *Select members*, search for and then select the desired service principal or managed identity, and then click *Select* to confirm.

Reviewer: Substeps? 
```suggestion
.. On the *Members* tab: 
... Choose an assignee by selecting the type of assignee: “User, group, or service principal” (Or “Managed identity”). 
... Click *Select members*. 
... Search for and then select the desired service principal or managed identity. 
... Click *Select* to confirm.
```

Reviewer: Not sure what this means; maybe it is clearer in the UI

> “User, group, or service principal” (Or “Managed identity”)

Reviewer: Also, should this be in mark up as UI elements? 

```suggestion
.. On the *Members* tab, choose an assignee by selecting the type of assignee: *User, group, or service principal* (Or *Managed identity*), click *Select members*, search for and then select the desired service principal or managed identity, and then click *Select* to confirm.
```

.. On the *Review + assign* tab, review the settings.

.. To finish the role assignment, click *Review + assign*.
+
[NOTE]
====
If you just want to use a specific storage account to provision the Azure File share, you can also obtain the Azure identity (service principal or managed identity) permission to access the storage account only with the similar steps.
====

> . Create an Azure File storage class using a similar configuration to the following:

Reviewer: ```suggestion
. Create an Azure File storage class by using a similar configuration to the following:
```

+
.Example Azure File storage class YAML file
[source, yaml]
----
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: <sc-name-cross-sub> <1>
mount options:
  - mfsymlinks
  - cache=strict
  - nosharesock
  - actimeo=30
parameters:
  subscriptionID: <xxxx-xxxx-xxxx-xxxx-xxxx> <2>
  resourceGroup: <resource group name> <3>
  storageAccount: <storage account> <4> 
  skuName: <skuName> <5>
provisioner: file.csi.azure.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
----
<1> Name of the storage class
<2> The cross account subscription ID
<3> The resource group name in cross account subscription

> <4> Storage account name, if you want to specify your own

Reviewer: To be consistent?
```suggestion
<4> The storage account name, if you want to specify your own
```

> <4> Storage account name, if you want to specify your own
> <5> Name of the SKU type

Reviewer: To be consistent?
```suggestion
<4> The storage account name, if you want to specify your own
<5> The name of the SKU type
```

> . Create a persistent volume claim (PVC) specifying the Azure File storage class that you created in the previous step using a similar configuration to the following:

Reviewer: Maybe a little clearer?
```suggestion
. Create a persistent volume claim (PVC) that specifies the Azure File storage class you created in the previous step by using a similar configuration to the following:
```

+
.Example PVC YAML file
[source, yaml]
----
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: <pvc-name> <1>
spec:
  storageClassName: <sc-name-cross-sub> <2>
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 5Gi
----

> <1> `<pvc-name>` is the name of the PVC.
> <2> `<sc-name-cross-sub>` is the name of the storage class that you created in the previous step.

Reviewer: Do we need the variables in the call-outs? 
```suggestion
<1> The name of the PVC.
<2> The name of the storage class that you created in the previous step.
```

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-provisioning-procedure_{context}"]
= Dynamic provisioning across subscriptions for Azure File

.Prerequisites

> * Installed Azure cluster with the service principal or managed identity as an Azure identity

Author: Done.

> * Access to another subscription in the tenant with the cluster  

Author: Done.

* Logged in to the Azure CLI

.Procedure
To use Azure File dynamic provisioning across subscriptions:

. Record the Azure identity (service principal or managed identity) by running the following applicable commands. The Azure identity is needed in a later step:
+
* If using the _service principal_ as the Azure identity when installing the cluster:
+
[source,terminal]
----
$ sp_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)

$ az ad sp show --id ${sp_id} --query displayName --output tsv
----
+
* If using the _managed identity_ as the Azure identity when installing the cluster:
+
[source,terminal]
----
$ mi_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)

$ az identity list --query "[?clientId=='${mi_id}'].{Name:name}" --output tsv
----

> . Obtain the Azure identity (service principal or managed identity) permission to access the resource group in another subscription where you want to provision the Azure File share:

Author: Done.

.. Log in to the Azure portal and on the left menu, click *Resource groups*.

> .. Choose the resource group to which you want to assign a role by clicking *resource group* -> *Access control (IAM)* -> *Role assignments* tab to view current assignments, and then click *Add* > *Add role assignment*.

Author: Done.

.. On the *Role* tab, choose the contributor role to assign, and then click *Next*. You can also create and choose your own role with required permission.

.. On the *Members* tab: 
... Choose an assignee by selecting the type of assignee: user, group, or service principal (or managed identity). 
... Click *Select members*. 
... Search for, and then select the desired service principal or managed identity. 
... Click *Select* to confirm.
.. On the *Review + assign* tab, review the settings.

.. To finish the role assignment, click *Review + assign*.

> +

Author: Done.

// Module included in the following assemblies:
//
// * storage/container_storage_interface/persistent_storage-csi-azure-file.adoc
//
:_mod-docs-content-type: PROCEDURE
[id="persistent-storage-csi-azure-file-cross-sub-dynamic-provisioning-procedure_{context}"]
= Dynamic provisioning across subscriptions for Azure File

.Prerequisites

> * Installed Azure cluster with the service principal or managed identity as an Azure identity in one subscription (call it Subscription A)

Author: Fixed.

> * Access to another subscription (call it Subscription B) in the same tenant with the cluster

Reviewer 2: same tenant "as" the cluster?

This subscription (B) should be the one with the storage IIUC, better call it out.

Reviewer 3: That would be better.

Author: Done.

$ az ad sp show --id ${sp_id} --query displayName --output tsv
----
+
* If using the _managed identity_ as the Azure identity when installing the cluster:
+
[source,terminal]
----
$ mi_id=$(oc -n openshift-cluster-csi-drivers get secret azure-file-credentials -o jsonpath='{.data.azure_client_id}' | base64 --decode)

$ az identity list --query "[?clientId=='${mi_id}'].{Name:name}" --output tsv
----

. Grant the Azure identity (service principal or managed identity) permission to access the resource group in another Subscription B where you want to provision the Azure File share by doing one of the following:

* Run the following Azure CLI command:
+
[source,terminal]
----
az role assignment create \
  --assignee <object-id-or-app-id> \
  --role <role-name> \
  --scope /subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.Storage/storageAccounts/<storage-account-name>
----
+
Where:
+
`<object-id-or-app-id>`: The service principal or managed identity that you obtained from the previous step, such as `sp_id` or `mi_id`.
+
`<role-name>`: Role name. Contributor or your own role with required permissions.
+

> `<subscription-id>`: Cross subscription B ID.

Reviewer 2: do we need to call it "cross", it's just subscription B ID isn't it?

Author: Yes. I've changed it.

+

> `<resource-group-name>`: Cross subscription B resource group name.

Author: Done.

... Click *Select members*. 
... Search for, and then select the desired service principal or managed identity recorded in the previous step. 
... Click *Select* to confirm.
.. On the *Review + assign* tab, review the settings.

.. To finish the role assignment, click *Review + assign*.
+
[NOTE]
====
If you only want to use a specific storage account to provision the Azure File share, you can also obtain the Azure identity (service principal or managed identity) permission to access the storage account by using similar steps.
====

. Create an Azure File storage class by using a similar configuration to the following:
+
.Example Azure File storage class YAML file
[source, yaml]
----
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: <sc-name-cross-sub> <1>
mount options:
  - mfsymlinks
  - cache=strict
  - nosharesock
  - actimeo=30
parameters:
  subscriptionID: <xxxx-xxxx-xxxx-xxxx-xxxx> <2>
  resourceGroup: <resource group name> <3>
  storageAccount: <storage account> <4> 
  skuName: <skuName> <5>
provisioner: file.csi.azure.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
----
<1> The name of the storage class

> <2> The cross account subscription ID

Reviewer 2: Same here i think we should avoid using the word cross in snippets this can be misleading.
"Call it subscription B ID" or "Subscription ID where the file share resides."

Author: Done.

> <3> The resource group name in cross account subscription

Reviewer 2: same here
